### ISSUE TITLE
Docker Swarm service discovery missing 'ancestor' filter

### ISSUE THREAD

**Issue by camo-f** [prometheus/prometheus#10081]
Hello,

[Prometheus documentation](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#dockerswarm_sd_config) states that Swarm targets can be filtered with [Docker Engine API](https://docs.docker.com/engine/api/v1.40/#tag/Task)

```
# Optional filters to limit the discovery process to a subset of available
# resources.
# The available filters are listed in the upstream documentation:
# https://docs.docker.com/engine/api/v1.40/#operation/ContainerList
[ filters:
  [ - name: <string>
      values: <string>, [...] ]
```

I tried to filter by ancestor (i.e. image), based on the referenced Docker documentation
```
- job_name: 'mysqld-exporter'
    dockerswarm_sd_configs:
    - host: http://dockerd:2375
      role: tasks
      filters:
	- name: ancestor
          values:
            - prom/mysqld-exporter
```

Promtool returns an empty result, despite the Swarm cluster having `mysqld-exporter` containers.
```
$ promtool check service-discovery /etc/prometheus/prometheus.yml mysqld-exporter
[]
```

Either the `ancestor` filter is not working, or the referenced documentation is wrong. I believe https://docs.docker.com/engine/api/v1.40/#tag/Task should be the correct one (at least for tasks).

Thanks

**Comment by LeviHarrison** [comment]
The specific filters aren't implemented by Prometheus itself, we only have the functionality to pass them to the API. That means the issue either falls on your configuration or the Docker engine. I'd recommend checking what API version you have (via `docker version`), the `ancestor` filter seems to have been added in `v1.23`.

**Comment by dginhoux** [comment]
Hi, 

Like you, i've lot of problems when using filters.
A single filter on name attributes, or desired-state, or label doesn't work.

I'm not alone : https://samizdat.dev/prometheus-scrap-all-docker-swarm-tasks/

**Comment by LeviHarrison** [comment]
Could you please provide the configuration that you're using?

And also, do all filters not work, or just this one? The issue implies that only the `ancestor` filter doesn't work, but if all don't maybe it's a problem with Prometheus.

**Comment by dginhoux** [comment]
Hi, 
I've hardly worked on my configuration.

I use the filters like explained on the documentation and as finded on google : 
```
  - job_name: docker-swarm-services
    dockerswarm_sd_configs:
      - host: tcp://socket:2375
        role: tasks
        refresh_interval: 30s
        filters:
          - name: name
            values: ["haproxy"]
```

After one hour of tries and tries..., i look at this page and the light has arrived ! 
https://prometheus.io/docs/guides/dockerswarm/


The principe is simple....  no filter used at query level.
Every services are pulled from the socket and prometheus simple exclude/include them, as you want.


After less than 30min, i've created a "labels powered scraping" : 
Labels used in service : 
>       labels:
>         prometheus_enable: "true"
>         prometheus_job: powerdns-authorative
>         prometheus_scheme: http
>         prometheus_port: 8081
>         prometheus_path: /metrics

Prometheus job configuration :
```
  - job_name: docker-swarm-services
    dockerswarm_sd_configs:
      - host: tcp://socket:2375
        role: tasks
        refresh_interval: 30s
    relabel_configs:
      # only keep services with prometheus_enable set to true
      - source_labels: ["__meta_dockerswarm_service_label_prometheus_enable"]
        regex: "true"
        action: keep
      # remove port if specified and set instance with address
      - source_labels: [__address__]
        regex: "([^:]+):\\d+"
        target_label: instance
      # remove port if specified and re-set address
      - source_labels: [__address__]
        # regex: ([^:]+)(:\d+)?;(\d+)
        regex: "([^:]+):\\d+"
        target_label: address
      # set instance with service name and task slot
      - source_labels: ["__meta_dockerswarm_service_name", "__meta_dockerswarm_task_slot"]
        regex: (.*)
        target_label: instance
        replacement: $1
        separator: "."
      # only keep services in running state
      - source_labels: ["__meta_dockerswarm_task_desired_state"]
        regex: running
        action: keep
      # Scraping metrics via a certain network only
      - source_labels: ["__meta_dockerswarm_network_name"]
        regex: monitoring-public
        action: keep
      # only keep services with prometheus_job set and set job variable
      - source_labels: [__meta_dockerswarm_service_label_prometheus_job]
        regex: .+
        action: keep
      - source_labels: [__meta_dockerswarm_service_label_prometheus_job]
        target_label: job
      # only keep services with prometheus_scheme set and set __scheme__ variable
      - source_labels: [__meta_dockerswarm_service_label_prometheus_scheme]
        regex: .+
        action: keep
      - source_labels: [__meta_dockerswarm_service_label_prometheus_scheme]
        target_label: __scheme__
      # only keep services with prometheus_port set and set prometheus_port variable
      - source_labels: [__meta_dockerswarm_service_label_prometheus_port]
        regex: .+
        action: keep
      - source_labels: [__meta_dockerswarm_service_label_prometheus_port]
        target_label: prometheus_port
      # only keep services with prometheus_path set and set __metrics_path__ variable
      - source_labels: [__meta_dockerswarm_service_label_prometheus_path]
        regex: .+
        action: keep
      - source_labels: [__meta_dockerswarm_service_label_prometheus_path]
        target_label: __metrics_path__
      # concatenate address with prometheus_port to re-set __address__ variable
      - source_labels: ["address", "prometheus_port"]
        # regex: "([^:]+):\\d+"
        # target_label: address
        # replacement: ${1}:${2}:${3}
        target_label: __address__
        action: replace
        separator: ":"

```

At this time, i've migrated more than 70 services, previously declared in `static_configs` or `dns_sd_configs` and lot of jobs (haproxy, nginx, etc....)

Now, i've just 8 jobs and i'm happy !!!

**Comment by camo-f** [comment]
> The specific filters aren't implemented by Prometheus itself, we only have the functionality to pass them to the API. That means the issue either falls on your configuration or the Docker engine. I'd recommend checking what API version you have (via `docker version`), the `ancestor` filter seems to have been added in `v1.23`.

Hi @LeviHarrison 

Sorry, I should have been more precise. I'm using Docker 20.10, and some filters are perfectly working, like `name` or `desired-state`. That's why I suggested the issue may not be with Prometheus, but rather with the documentation referencing the wrong API (https://docs.docker.com/engine/api/v1.40/#operation/ContainerList vs https://docs.docker.com/engine/api/v1.40/#tag/Task). In the second one, there is no `ancestor` filter, so Prometheus is working as expected.

**Comment by LeviHarrison** [comment]
> That's why I suggested the issue may not be with Prometheus, but rather with the documentation referencing the wrong API

Ah, I should have read more carefully, apologies. I believe you're right. There should be links to the `ServiceList`, `TaskList`, and `NodeList` API documentation, as those are what the Swarm SD uses. Interestingly enough, those links used to be there, but were replaced with a link to the `ContainerList` documentation in #8629 (https://github.com/prometheus/prometheus/commit/b253056163b705a13a4e2f961348717ae9657824#diff-17f1012e0c2fbd9bcd8dff3c23b18ff4b6676eef3beca6f8a3e72e6a36633334L604-L606). I wonder why.

cc @roidelapluie @xperimental

**Comment by xperimental** [comment]
Thanks for the highlight. It looks like I moved around the documentation too much and added the wrong links to both sections (Docker Swarm referencing Docker Engine and vice-versa).

I'll give this another read and then create a PR fixing the documentation. Am I understanding the discussion correctly, that the feature works properly only the documentation is affected or is it also broken?

**Comment by LeviHarrison** [comment]
Thanks for the quick response and the explanation, I really appreciate it. Prometheus is working properly, just the documentation is inaccurate.

**Comment by LeviHarrison** [comment]
Documentation was fixed #10252
